plugins/inputs/tencentcloudcm/discover.go

```diff
@@ -0,0 +1,193 @@
+	for namespace, p := range d.registry {
+		if d.DiscoveredMetrics == nil {
+			d.DiscoveredMetrics = map[string][]string{}
+		}
+		d.DiscoveredMetrics[namespace] = p.Metrics()
+	}
+}
+
+func (d *discoverTool) discoverObjects(accounts []*Account, endpoint string) {
+	discoveredObjects := map[string]discoverObject{}
+	for _, account := range accounts {
+		for _, namespace := range account.Namespaces {
+			for _, region := range namespace.Regions {
+				// skip discover if instances are explicitly specified
+				if len(region.monitorInstances) != 0 {
+					continue
+				}
+				p, ok := d.registry[namespace.Name]
+				if !ok {
+					continue
+				}
+				discoveredObject, err := discover(account.crs, region.RegionName, endpoint, p, d.Log)
+				if err != nil {
+					d.Log.Errorf("discover account:%s region:%s endpoint:%s failed, error: %s",
+						account.Name, region.RegionName, endpoint, err)
+					continue
+				}
+				discoveredObjects[newKey(account.Name, namespace.Name, region.RegionName)] = discoveredObject
+			}
+		}
+	}
+	d.rw.Lock()
+	d.DiscoveredObjects = discoveredObjects
+	d.rw.Unlock()
+}
+
+// Discover discovers instances of registered products
+func (d *discoverTool) Discover(accounts []*Account, interval config.Duration, endpoint string) {
+	ticker := time.NewTicker(time.Duration(interval))
+	defer ticker.Stop()
+
+	d.discoverObjects(accounts, endpoint)
+
+	// discover instances periodically
+	for range ticker.C {
+		d.discoverObjects(accounts, endpoint)
+	}
+}
+
+// GetInstance Get discovered instance detail
+func (d *discoverTool) GetInstance(account, namespace, region, key string) map[string]string {
+	d.rw.RLock()
+	defer d.rw.RUnlock()
+	discoverObject, ok := d.DiscoveredObjects[newKey(account, namespace, region)]
+	if !ok {
+		return nil
+	}
+	return discoverObject.Instances[key]
+}
+
+// GetMonitorInstances Get discovered monitor instances
+func (d *discoverTool) GetMonitorInstances(account, namespace, region string) []*monitor.Instance {
+	d.rw.RLock()
+	defer d.rw.RUnlock()
+	discoverObject, ok := d.DiscoveredObjects[newKey(account, namespace, region)]
+	if !ok {
+		return nil
+	}
+	return discoverObject.MonitorInstances
+}
+
+func discover(crs *common.Credential, region, endpoint string, p Product, l telegraf.Logger) (discoverObject, error) {
+	discoverObject := discoverObject{
+		Instances: map[string]map[string]string{},
+	}
+
+	const offset, limit = int64(0), int64(100)
+	instances := []map[string]string{}
+
+	total, instancesByPage, err := p.Discover(crs, region, endpoint, offset, limit)
+	if err != nil {
+		return discoverObject, err
+	}
+	instances = append(instances, instancesByPage...)
+
+	// discover all instances if total count is bigger than limit
+	for i := 1; i < int(int64(total)/limit)+1; i++ {
+		_, instancesByPage, err := p.Discover(crs, region, endpoint, offset+(int64(i)*limit), limit)
+		if err != nil {
+			return discoverObject, err
+		}
+		instances = append(instances, instancesByPage...)
+	}
```

Comment:
Why not do
```suggestion
	var offset int64
	instances := []map[string]string{}

	total, instancesByPage, err := p.Discover(crs, region, endpoint, offset, limit)
	if err != nil {
		return discoverObject, err
	}
	instances = append(instances, instancesByPage...)
	offset += limit

	// discover all instances if total count is bigger than limit
	for i := int64(1); i < int64(total-1)/limit+1; i++ {
		_, instancesByPage, err := p.Discover(crs, region, endpoint, offset, limit)
		if err != nil {
			return discoverObject, err
		}
		instances = append(instances, instancesByPage...)
		offset += limit
	}
```
which makes the code more readable. Please not that I don't define `total` in this function, but you can define it as a constant on the top of the file or pass it as parameter that can be user-defined to allow larger/smaller batches.